shared/diagram.py

```diff
@@ -45,14 +133,102 @@ def make_directories():
+        return ordered_resources
+
+    def process_relationships(
+        self,
+        grouped_resources: Dict[str, List[Resource]],
+        resource_relations: List[ResourceEdge],
+    ) -> List[ResourceEdge]:
+        return resource_relations
+
     @exception
-    def generate_diagram(self, resources: List[Resource]):
-        raise NotImplementedError("Implement the diagram generation logic")
+    def generate_diagram(
+        self, resources: List[Resource], initial_resource_relations: List[ResourceEdge]
+    ):
+        ordered_resources = self.group_by_group(resources, initial_resource_relations)
+        relations = self.process_relationships(
+            ordered_resources, initial_resource_relations
+        )
+
+        with Diagram(
+            name=self.name,
+            filename=PATH_DIAGRAM_OUTPUT + self.filename,
+            direction="TB",
+            graph_attr={"nodesep": "2.0", "ranksep": "1.0", "splines": "curved"},
+        ) as d:
+            d.dot.engine = self.engine
+
+            self.draw_diagram(ordered_resources=ordered_resources, relations=relations)
+
+    def draw_diagram(self, ordered_resources, relations):
+        already_drawn_elements = {}
+
+        # Import all AWS nodes
+        for module in Mapsources.diagrams_modules:
+            exec("from diagrams.aws." + module + " import *")
+
+        nodes: Dict[ResourceDigest, any] = {}
+        # Iterate resources to draw it
+        for group_name in ordered_resources:
+            if group_name == "":
+                for resource in ordered_resources[group_name]:
+                    node = eval(Mapsources.mapresources.get(resource.digest.type))(
+                        resource.name
+                    )
+                    nodes[resource.digest] = node
+            else:
+                with Cluster(group_name.capitalize() + " resources") as cluster:
+                    nodes[ResourceDigest(id=group_name, type=DIAGRAM_CLUSTER)] = cluster
+                    for resource in ordered_resources[group_name]:
+                        node = eval(Mapsources.mapresources.get(resource.digest.type))(
+                            resource.name
+                        )
+                        nodes[resource.digest] = node
+
+        for resource_relation in relations:
+            if (
+                resource_relation.from_node in nodes
+                and resource_relation.to_node in nodes
+            ):
+                from_node = nodes[resource_relation.from_node]
+                to_node = nodes[resource_relation.to_node]
+                if resource_relation.from_node not in already_drawn_elements:
+                    already_drawn_elements[resource_relation.from_node] = {}
+                if (
+                    resource_relation.to_node
+                    not in already_drawn_elements[resource_relation.from_node]
+                ):
+                    from_node >> Edge(label=resource_relation.label) >> to_node
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Expression "((from_node) >> (Edge(label=resource_relation.label))) >> (to_node)" is assigned to nothing](https://app.codacy.com/gh/Cloud-Architects/cloud-discovery/pullRequest?prid=5655786)